src/compiler/scala/tools/nsc/ast/parser/Parsers.scala

```diff
@@ -1099,7 +1101,14 @@ self =>
           else
             mkOp(infixType(InfixMode.RightOp))
         }
+        def isNegatedLiteralType = settings.YliteralTypes && (
+          t match { // the token for `t` (Ident("-")) has already been read, thus `isLiteral` below is looking at next token (must be a literal)
+            case Ident(name) if isLiteral => name == nme.MINUS.toTypeName
+            case _ => false
+          }
+        )
         if (isIdent) checkRepeatedParam orElse asInfix
+        else if (isNegatedLiteralType) atPos(t.pos.start){SingletonTypeTree(literal(isNegated = true, start = t.pos.start))}
```

Comment:
It would be good to include a test that covers all the parser changes and that runs with `-Yrange-pos` enabled. We can also eyeball the output of `-Yrange-pos -Xprint:parser -Xprint-pos` for that file to check that positions are precise.